### PR TITLE
Updates to OAuth samples

### PR DESCRIPTION
**Pull request by ZackAllen**
This PR aims to fix several issues with OAuth in the samples.
- Offline map samples use an OAuth pattern where API key is sufficient.
- Create and save map, Search for webmap, Authenticate with OAuth, and Navigate in AR all use repeat OAuth code. It is better if each platform has its own helper class for authentication.
- Authentication itself is using the wrong `TokenAuthenticationType` in some cases.
- There are several bugs in OAuth samples (especially when a login is cancelled).

PR does the following:
- Creates helper classes for each platforms AGOL OAuth authentication.
- Updates viewers to display source code for the helper classes.
- Updates viewer code to properly clear the authentication manager.
- Removes all OAuth code from Offline map samples.
- Updates "Create and save map" to use new Basemap constructors.
- Replaces Xamarin.Auth with Xamarin.Essentials.
- Corrects wrong usage of OAuthImplicit.

### CODE REVIEW

File: src/Forms/iOS/ArcGISRuntime.Xamarin.Forms.iOS.csproj [hunk 1]
```diff
@@ -37,6 +38,8 @@
     <MtouchI18n />
     <MtouchHttpClientHandler>NSUrlSessionHandler</MtouchHttpClientHandler>
     <LangVersion>7</LangVersion>
+    <CodesignKey>Apple Development: Zack Allen (H3TP7QX4B5)</CodesignKey>
```
Comment (mbranscomb) [inline src/Forms/iOS/ArcGISRuntime.Xamarin.Forms.iOS.csproj:41]: Is this intentional?

File: src/Forms/iOS/ArcGISRuntime.Xamarin.Forms.iOS.csproj [hunk 2]
```diff
@@ -73,7 +76,7 @@
     <WarningLevel>4</WarningLevel>
     <ConsolePause>false</ConsolePause>
     <MtouchArch>ARM64</MtouchArch>
-    <CodesignKey>iPhone Developer</CodesignKey>
+    <CodesignKey>Apple Distribution: ESRI (75Z67E6SV2)</CodesignKey>
```
Comment (mbranscomb) [inline src/Forms/iOS/ArcGISRuntime.Xamarin.Forms.iOS.csproj:79]: Probably remove this?

File: src/Forms/iOS/ArcGISRuntime.Xamarin.Forms.iOS.csproj [hunk 3]
```diff
@@ -88,6 +91,7 @@
     <MtouchI18n />
     <LangVersion>7</LangVersion>
     <MtouchHttpClientHandler>NSUrlSessionHandler</MtouchHttpClientHandler>
+    <CodesignProvision>com.esri.* Wildcard Distribution</CodesignProvision>
```
Comment (mbranscomb) [inline src/Forms/iOS/ArcGISRuntime.Xamarin.Forms.iOS.csproj:94]: Intentional?

File: src/UWP/ArcGISRuntime.UWP.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            // If a client secret has been configured, set the authentication type to OAuth client credentials.
+            if (!string.IsNullOrEmpty(ClientSecret))
+            {
+                // Use OAuthClientCredentials if you need a refresh token (and have specified a valid client secret).
+                portalServerInfo.TokenAuthenticationType = TokenAuthenticationType.OAuthClientCredentials;
+                portalServerInfo.OAuthClientInfo.ClientSecret = ClientSecret;
+            }
+
+            // Get a reference to the (singleton) AuthenticationManager for the app
+            AuthenticationManager thisAuthenticationManager = AuthenticationManager.Current;
```
Comment (mbranscomb) [inline src/UWP/ArcGISRuntime.UWP.Viewer/Helpers/ArcGISLoginPrompt.cs:102]: I'd remove this in favor of using `AuthenticationManager.Current` directly below.

File: src/UWP/ArcGISRuntime.UWP.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+            // Register the ArcGIS Online server information with the AuthenticationManager
+            thisAuthenticationManager.RegisterServer(portalServerInfo);
```
Comment (mbranscomb) [inline src/UWP/ArcGISRuntime.UWP.Viewer/Helpers/ArcGISLoginPrompt.cs:105]: Use `AuthenticationManager.Current`?

File: src/UWP/ArcGISRuntime.UWP.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+            // Create a new ChallengeHandler that uses a method in this class to challenge for credentials
+            thisAuthenticationManager.ChallengeHandler = new ChallengeHandler(PromptCredentialAsync);
```
Comment (mbranscomb) [inline src/UWP/ArcGISRuntime.UWP.Viewer/Helpers/ArcGISLoginPrompt.cs:108]: Use `AuthenticationManager.Current`?

File: src/UWP/ArcGISRuntime.UWP.Viewer/Samples/Map/AuthorMap/AuthorMap.xaml.cs [hunk 7]
```diff
@@ -63,22 +53,30 @@ public AuthorMap()
...
-            MyMapView.Map = new Map();
+            ArcGISLoginPrompt.SetChallengeHandler();
+
+            bool loggedIn = await ArcGISLoginPrompt.EnsureAGOLCredentialAsync();
+
+            // Show a plain gray map in the map view
+            if (loggedIn)
+            {
+                MyMapView.Map = new Map(BasemapStyle.ArcGISLightGray);
+                _ = MyMapView.Map.Basemap.LoadAsync();
```
Comment (mbranscomb) [inline src/UWP/ArcGISRuntime.UWP.Viewer/Samples/Map/AuthorMap/AuthorMap.xaml.cs:69]: Is loading the basemap specifically the same pattern used in the AuthorMap samples on the other .NET platforms?

File: src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        public static async Task<bool> EnsureAGOLCredentialAsync()
+        {
+            bool loggedIn = false;
+
+            try
+            {
+                // Create a challenge request for portal credentials (OAuth credential request for arcgis.com)
+                CredentialRequestInfo challengeRequest = new CredentialRequestInfo
+                {
+                    // Use the OAuth implicit grant flow
```
Comment (mbranscomb) [inline src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs:43]: OAuthAuthorizationCode

File: src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+            ServerInfo portalServerInfo = new ServerInfo(new Uri(ArcGISOnlineUrl))
+            {
+                TokenAuthenticationType = TokenAuthenticationType.OAuthAuthorizationCode,
+                OAuthClientInfo = new OAuthClientInfo(AppClientId, new Uri(OAuthRedirectUrl))
+            };
+
+            // If a client secret has been configured, set the authentication type to OAuth client credentials.
+            if (!string.IsNullOrEmpty(ClientSecret))
+            {
+                // Use OAuthClientCredentials if you need a refresh token (and have specified a valid client secret).
```
Comment (mbranscomb) [inline src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs:100]:
```suggestion
                // Use OAuthClientCredentials if you have specified a client secret.
```

File: src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+                portalServerInfo.TokenAuthenticationType = TokenAuthenticationType.OAuthClientCredentials;
+                portalServerInfo.OAuthClientInfo.ClientSecret = ClientSecret;
+            }
+
+            // Get a reference to the (singleton) AuthenticationManager for the app
+            AuthenticationManager thisAuthenticationManager = AuthenticationManager.Current;
```
Comment (mbranscomb) [inline src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs:106]: Use `AuthenticationManager.Current` directly below.

File: src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+
+            // Register the ArcGIS Online server information with the AuthenticationManager
+            thisAuthenticationManager.RegisterServer(portalServerInfo);
```
Comment (mbranscomb) [inline src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs:109]: Use `AuthenticationManager.Current` directly

File: src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+
+            // Use the OAuthAuthorize class in this project to create a new web view to show the login UI
+            thisAuthenticationManager.OAuthAuthorizeHandler = new OAuthAuthorize();
```
Comment (mbranscomb) [inline src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs:112]: Use `AuthenticationManager.Current` directly

File: src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+
+            // Create a new ChallengeHandler that uses a method in this class to challenge for credentials
+            thisAuthenticationManager.ChallengeHandler = new ChallengeHandler(PromptCredentialAsync);
```
Comment (mbranscomb) [inline src/WPF/ArcGISRuntime.WPF.Viewer/Helpers/ArcGISLoginPrompt.cs:115]: Use `AuthenticationManager.Current` directly

File: src/WPF/ArcGISRuntime.WPF.Viewer/Samples/Map/AuthorMap/AuthorMap.xaml.cs [hunk 14]
```diff
@@ -65,27 +53,32 @@ public AuthorMap()
...
+            ArcGISLoginPrompt.SetChallengeHandler();
+
+            bool loggedIn = await ArcGISLoginPrompt.EnsureAGOLCredentialAsync();
+
             // Show a plain gray map in the map view
-            MyMapView.Map = new Map(Basemap.CreateLightGrayCanvas());
+            if (loggedIn)
+            {
+                MyMapView.Map = new Map(BasemapStyle.ArcGISLightGray);
+                _ = MyMapView.Map.Basemap.LoadAsync();
```
Comment (mbranscomb) [inline src/WPF/ArcGISRuntime.WPF.Viewer/Samples/Map/AuthorMap/AuthorMap.xaml.cs:69]: Await the LoadAsync.

File: src/iOS/Xamarin.iOS/Helpers/ArcGISLoginPrompt.cs [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                Credential cred = await AuthenticationManager.Current.GetCredentialAsync(challengeRequest, false);
+                loggedIn = cred != null;
+            }
+            catch (OperationCanceledException)
+            {
+                // OAuth login was canceled, no need to display error to user.
+            }
+            catch (Exception ex)
+            {
+                //await Application.Current.MainPage.DisplayAlert("Login failed", ex.Message, "OK");
```
Comment (mbranscomb) [inline src/iOS/Xamarin.iOS/Helpers/ArcGISLoginPrompt.cs:67]: Is this intentionally commented out?

File: src/iOS/Xamarin.iOS/Helpers/ArcGISLoginPrompt.cs [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+            ServerInfo portalServerInfo = new ServerInfo(new Uri(ArcGISOnlineUrl))
+            {
+                TokenAuthenticationType = TokenAuthenticationType.OAuthAuthorizationCode,
+                OAuthClientInfo = new OAuthClientInfo(AppClientId, new Uri(OAuthRedirectUrl))
+            };
+
+            // If a client secret has been configured, set the authentication type to OAuth client credentials.
+            if (!string.IsNullOrEmpty(ClientSecret))
+            {
+                // Use OAuthClientCredentials if you need a refresh token (and have specified a valid client secret).
```
Comment (mbranscomb) [inline src/iOS/Xamarin.iOS/Helpers/ArcGISLoginPrompt.cs:85]:
```suggestion
                // Use OAuthClientCredentials if you have specified a client secret.
```

File: src/iOS/Xamarin.iOS/Samples/Map/AuthorMap/AuthorMap.cs [hunk 17]
```diff
@@ -69,31 +70,27 @@ public class AuthorMap : UIViewController, IOAuthAuthorizeHandler
...
-            _myMapView.Map = new Map(Basemap.CreateLightGrayCanvas());
+            // Prompt the user for a login. (Needed for loading basemaps without an API key.)
+            ArcGISLoginPrompt.SetChallengeHandler(this);
+            bool loggedIn = await ArcGISLoginPrompt.EnsureAGOLCredentialAsync();
+
+            // Show a plain gray map in the map view
+            if (loggedIn)
+            {
+                _myMapView.Map = new Map(BasemapStyle.ArcGISLightGray);
+                _ = _myMapView.Map.Basemap.LoadAsync();
```
Comment (mbranscomb) [inline src/iOS/Xamarin.iOS/Samples/Map/AuthorMap/AuthorMap.cs:91]: Await the LoadAsync

What happens if it returns an exception?

**Review by mbranscomb — CHANGES_REQUESTED** [review]
@ZackAllen Wow, this is a massive PR - overall nice work.

- Requested/suggested a few changes.
- Nice to see async voids being gradually replaced with async tasks where appropriate. I noticed though that not all are updated in this PR, which is OK, as long as there's a plan to address the others in the future.

File: src/Android/Xamarin.Android/Samples/Map/SearchPortalMaps/SearchPortalMaps.cs [hunk 18]
```diff
@@ -68,22 +53,32 @@ protected override void OnCreate(Bundle bundle)
...
-            // Create a new Map instance
-            Map myMap = new Map(BasemapStyle.ArcGISLightGray);
+            // Remove API key.
+            ApiKeyManager.DisableKey();
+
+            ArcGISLoginPrompt.SetChallengeHandler();
 
-            // Provide default Map to the MapView
-            _myMapView.Map = myMap;
+            bool loggedIn = await ArcGISLoginPrompt.EnsureAGOLCredentialAsync(this);
```
[the remark this answers is not in the document]
Comment (ZackAllen) [reply src/Android/Xamarin.Android/Samples/Map/SearchPortalMaps/SearchPortalMaps.cs:68]: `EnsureAGOLCredentialAsync` catches all potential exceptions. Will always return a bool.

**Comment by ZackAllen** [comment]
The author map basemap loads were an artifact from when I was trying to rewrite the sample by setting the API key on the basemap. They are not needed for the new pattern.

**Comment by cgutierrez-gis** [comment]
Some issues I'm seeing in the impacted samples:

(all sample viewers)
Create and save map - Selecting to create a new map wipes out map view and does not repopulate the map view after a selection is made.

**Review by nCastle1 — CHANGES_REQUESTED** [review]
I noticed on WPF that when switching to a sample with OAuth, the OAuth dialog is shown before the new sample's UI is shown. Ideally the old sample would be hidden and replaced by the new sample, or at least replaced with a loading screen, before the OAuth dialog appears.

The iOS samples have a lot of unused `OAuth2Authenticator` member variables, I'd like to see those cleaned up.

Everything else looks good.

**Review by nCastle1 — APPROVED** [review]

**Review by mbranscomb — APPROVED** [review]

**Comment by mbranscomb** [comment]
@ZackAllen Looks like you resolved all my comments/questions. 

A good and necessary enhancement. 

Approved.

**Pull request merged**